rust/sdk/examples/chat/cmdline.rs

```diff
@@ -12,40 +12,138 @@ pub enum Cmdline {
+        Self {
+            edit: Some(mk_identity_text_area(&me)),
+            val: mk_identity_text_area(&me),
+        }
+    }
+}
+
 pub async fn cmdline(
     mut cell: ActoCell<Cmdline, impl ActoRuntime>,
     display: ActoRef<Display>,
     mut messages: ActoRef<Messages>,
 ) {
-    let text_area = Writer::new(mk_text_area());
+    let identity = Writer::new(Identity::default());
+    let text_area = Writer::new(mk_message_text_area());
+    display.send(Display::UpdateIdentity(identity.reader()));
     display.send(Display::Cmdline(text_area.reader()));
 
     while let ActoInput::Message(msg) = cell.recv().await {
         match msg {
             Cmdline::Event(event) => {
-                if let Event::Key(key) = event {
-                    // FIXME modifiers are not recognised, so no ctrl-enter
-                    if key.code == KeyCode::Enter {
-                        let mut text_area = text_area.write();
-                        let text = std::mem::replace(&mut *text_area, mk_text_area())
-                            .into_lines()
-                            .join("\n");
-                        tracing::info!("publishing message: {}", text);
-                        messages.send(Messages::Publish(messages::Event::new("me".to_owned(), text)));
-                    } else {
-                        text_area.write().input(key);
+                // Prevent double input because of KeyEventKind::Release
+                let key = match event {
+                    Event::Key(key) if key.kind == KeyEventKind::Press || key.kind == KeyEventKind::Repeat => key,
+                    _ => continue,
+                };
+
+                let is_editing_identity = { (identity.read()).edit.is_some() };
+                match is_editing_identity {
+                    // When editing identity
+                    true => {
+                        {
+                            let mut identity = identity.write();
+                            match key.code {
+                                KeyCode::Esc => {
+                                    identity.edit = None;
+                                }
+                                KeyCode::Enter => {
+                                    if let Some(identity_buffer_value) = identity.edit.take() {
+                                        identity.val = identity_buffer_value;
+                                    }
+                                }
+                                _ => {
+                                    if let Some(edit) = &mut identity.edit {
+                                        edit.input(key);
+                                    }
+                                }
+                            }
+                        }
+                        display.send(Display::UpdateIdentity(identity.reader()));
+                    }
+                    // When chatting
+                    false => {
+                        match key.code {
```

Comment:
I am not sure about spawning another actor because then we will indicate that somehow identity controller can potentially move non-synchronously with the rest of the cmdline.
I am also not sure about `input` recognizing `page up` and `page down` because at least the model in my head says precisely what you have written: "Input is alive and running unless there is a CTRL+C"

However, I think it is good to separate the "identity" mode as a "sub-state-machine" of cmdline's state machine. But this is done for code organization and not software architecture.

Reply:
The identity edit field is not used concurrently with the message composer (which I somehow misnamed `cmdline`), either one has focus or the other. In the future we may have other things focused, so I do think we need to have a focus handler. This handler cannot be inside `cmdline` because `cmdline` can go out of focus itself, hence my plan to place this code inside `input` — after all, it is about routing input events to the right places.

I also think PageUp/PageDown have nothing to do with `cmdline`, they are only of interest to `input` and `display`.